File: 2058.find-the-minimum-and-maximum-number-of-nodes-between-critical-points.py
#
# @lc app=leetcode id=2058 lang=python3
#
# [2058] Find the Minimum and Maximum Number of Nodes Between Critical Points
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def nodesBetweenCriticalPoints(self, head: Optional[ListNode]) -> List[int]:
        positions = []
        prev, curr = head, head.next
        i = 1
        while curr and curr.next:
            if (curr.val > prev.val and curr.val > curr.next.val) or (curr.val < prev.val and curr.val < curr.next.val):
                positions.append(i)
            prev = curr
            curr = curr.next
            i += 1
        # Verification step: check if enough critical points were found
        if len(positions) < 2:
            return [-1, -1]
        # Calculate min and max distances
        min_dist = min(b - a for a, b in zip(positions, positions[1:]))
        max_dist = positions[-1] - positions[0]
        # Final check to ensure distances are valid
        if min_dist <= 0 or max_dist <= 0:
            return [-1, -1]
        return [min_dist, max_dist]
# @lc code=end

File: 2058.find-the-minimum-and-maximum-number-of-nodes-between-critical-points.java
#
# @lc app=leetcode id=2058 lang=java
#
# [2058] Find the Minimum and Maximum Number of Nodes Between Critical Points
#
# @lc code=start
/**
* Definition for singly-linked list.
* public class ListNode {
*     int val;
*     ListNode next;
*     ListNode() {}
*     ListNode(int val) { this.val = val; }
*     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
* }
*/
class Solution {
    public int[] nodesBetweenCriticalPoints(ListNode head) {
        List<Integer> values = new ArrayList<>();
        ListNode node = head;
        while (node != null) {
            values.add(node.val);
            node = node.next;
        }
        List<Integer> criticalIndices = new ArrayList<>();
        for (int i = 1; i < values.size() - 1; ++i) {
            int prev = values.get(i - 1);
            int curr = values.get(i);
            int next = values.get(i + 1);
            if ((curr > prev && curr > next) || (curr < prev && curr < next)) {
                criticalIndices.add(i);
            }
        }
        // Verification step: all indices should be in [1, values.size() - 2]
        for (int idx : criticalIndices) {
            if (idx <= 0 || idx >= values.size() - 1) {
                return new int[]{-1, -1};
            }
        }
        if (criticalIndices.size() < 2) return new int[]{-1, -1};
        int minDist = Integer.MAX_VALUE;
        for (int i = 1; i < criticalIndices.size(); ++i) {
            minDist = Math.min(minDist, criticalIndices.get(i) - criticalIndices.get(i - 1));
        }
        int maxDist = criticalIndices.get(criticalIndices.size() - 1) - criticalIndices.get(0);
        return new int[]{minDist, maxDist};
    }
}
# @lc code=end